src/chat.rs

```diff
@@ -435,6 +435,8 @@ impl ChatId {
             return Ok(());
         }
 
+        let msg_is_some = msg.is_some();
```

Comment:
You could do:
```diff
diff --git a/src/chat.rs b/src/chat.rs
index d60fc0e4..61b1fa9c 100644
--- a/src/chat.rs
+++ b/src/chat.rs
@@ -429,15 +429,13 @@ pub async fn delete(self, context: &Context) -> Result<(), Error> {
     pub async fn set_draft(
         self,
         context: &Context,
-        msg: Option<&mut Message>,
+        mut msg: Option<&mut Message>,
     ) -> Result<(), Error> {
         if self.is_special() {
             return Ok(());
         }
 
-        let msg_is_some = msg.is_some();
-
-        let changed = match msg {
+        let changed = match &mut msg {
             None => self.maybe_delete_draft(context).await?,
             Some(msg) => self.set_draft_raw(context, msg).await?,
         };
@@ -445,7 +443,7 @@ pub async fn set_draft(
         if changed {
             context.emit_event(EventType::MsgsChanged {
                 chat_id: self,
-                msg_id: if msg_is_some {
+                msg_id: if msg.is_some() {
                     match self.get_draft_msg_id(context).await? {
                         Some(msg_id) => msg_id,
                         None => MsgId::new(0),

```

The thing is, if you do `match msg`, then it (the option) will be consumed. 

If you do `match &msg`, then it's behind a shared reference and can't be borrowed mutably for `set_draft_raw()`.

So, you have to `match &mut msg`. But in order to mutably borrow `msg`, the variable `msg` has to be mutable, so we have to declare it as `mut` in the parameter list. Note that this doesn't change the function signature, it just means that the value of the local `msg` variable could change. Which is true, `do_set_draft()` sometimes alters `msg.param`, so if we called `msg.map(|m| m.param)`, instead of `msg.is_some()`, then we would get another result depending on whether we call it at before or after calling `set_draft_raw()`.

Sometimes it's just nice to write down why the borrowing rules make sense :)

Reply:
thanks now I understand it better

Reply:
I have recommended it on IRC once I think, relevant: https://www.possiblerust.com/guide/how-to-read-rust-functions-part-1